apm-agent-plugins/apm-struts-plugin/src/main/java/co/elastic/apm/agent/struts/Struts2TransactionNameInstrumentation.java

```diff
@@ -0,0 +1,76 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2021 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.struts;
+
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.stacktrace.StacktraceConfiguration;
+import co.elastic.apm.agent.sdk.ElasticApmInstrumentation;
+import net.bytebuddy.description.NamedElement;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import net.bytebuddy.matcher.ElementMatchers;
+
+import java.util.Collection;
+
+import static co.elastic.apm.agent.bci.bytebuddy.CustomElementMatchers.isInAnyPackage;
+import static java.util.Collections.singletonList;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+public class Struts2TransactionNameInstrumentation extends ElasticApmInstrumentation {
+
+    private final StacktraceConfiguration config;
+
+    public Struts2TransactionNameInstrumentation(ElasticApmTracer tracer) {
+        config = tracer.getConfig(StacktraceConfiguration.class);
+    }
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        return named("org.apache.struts2.factory.StrutsActionProxy")
+            .or(isInAnyPackage(config.getApplicationPackages(), ElementMatchers.<NamedElement>any()));
```

Comment:
That's a really expensive matcher. Especially because the application_packages setting is optional and this matcher runs even if struts is not on the classpath.

The pre-filter is intended to return a fast executing filter that may result in false negatives. For example `nameContains("ActionProxy")` limits the number of type descriptions Byte Buddy has to load from disk considerably and there's a good chance that custom implementations are still matched. There's a flag to turn off all pre-matchers in case they exclude classes that should be implemented.

Reply:
I am fine with `nameContains("ActionProxy")`. Should I change it?

Reply:
Let's see what @SylvainJuge says but I'd say yes.

Reply:
Yes, I agree with @felixbarny here, we could simply have it with `nameContains("ActionProxy")` in `getTypeMatcherPreFilter` and avoid relying on `application_packages`:
- there are only a few cases where `StrutsActionProxy` class is extended (heuristic being search on public code samples in github).
- `application_packages` is often not set or set improperly, thus avoiding relying on it is better.

Reply:
Done